File: rust/src/model/post.rs
use crate::grpc::post::{SubmitRentRequest, SubmitHelpRequest,};
use sqlx::FromRow;
use tonic::Request;

#[derive(Debug, FromRow)]
pub struct RentPost {
    pub id: i64,
    pub name: String,
    pub phone: String,
    pub room_addr: String,
    pub room_area: i32,
    pub room_type: String,
    pub room_orientation: String,
    pub room_floor: i32,
    pub description: String,
    pub price: i32,
    pub restriction: String,
    pub create_by: String,
    pub uuid: String,
    pub release_time: String,
    pub pictures: Vec<Vec<u8>>,
}

impl RentPost {
    pub fn text(&self) -> (String, String) {
        (self.uuid.clone(), vec![self.name.as_str(), self.phone.as_str(), self.room_addr.as_str(), self.room_area.to_string().as_str(),
        self.room_type.as_str(), self.room_orientation.as_str(), self.description.as_str(), self.price.to_string().as_str(),
        self.release_time.as_str()].join(" "))
    }
}

impl From<Request<SubmitRentRequest>> for RentPost {
    fn from(submit_rent_request: Request<SubmitRentRequest>) -> Self {
        let request = submit_rent_request.into_inner();
        RentPost {
            id: 0i64,
            name: request.name,
            phone: request.phone,
            room_addr: request.room_addr,
            room_area: request.room_area,
            room_type: request.room_type,
            room_orientation: request.room_orientation,
            room_floor: request.room_floor,
            description: request.description,
            price: request.price,
            restriction: request.restriction,
            create_by: request.create_by,
            uuid: request.uuid,
            release_time: request.release_time,
            pictures: request.pictures,
        }
    }
}

#[derive(Debug, FromRow)]
pub struct HelpPost {
    id: i64,
    pub name: String,
    pub phone: String,
    pub expected_addr: String,
    pub expected_price: i32,
    pub demands: String,
    pub create_by: String,
    pub uuid: String,
    pub release_time: String,
}

impl HelpPost {
    pub fn text(&self) -> (String, String) {
        (self.uuid.clone(), vec![self.name.as_str(), self.phone.as_str(), self.expected_addr.as_str(),
        self.expected_price.to_string().as_str(), self.demands.to_string().as_str(), self.release_time.as_str()].join(" "))
    }
}

impl From<Request<SubmitHelpRequest>> for HelpPost {
    fn from(submit_help_request: Request<SubmitHelpRequest>) -> Self {
        let request = submit_help_request.into_inner();
        HelpPost {
            id: 0i64,
            name: request.name,
            phone: request.phone,
            expected_addr: request.expected_addr,
            expected_price: request.expected_price,
            demands: request.demands,
            create_by: request.create_by,
            uuid: request.uuid,
            release_time: request.release_time,
        }
    }
}

File: admin/lib/model/post.dart
import 'dart:typed_data';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

abstract class Post {
  String uuid;
  String name;
  String phone;
  String releaseTime;

  Post(this.uuid, this.name, this.phone, this.releaseTime);

  Widget buildPanel(BuildContext context,
      {required void Function()? onVerified});
}

class HelpPost extends Post {
  String? expectedAddr;
  int? expectedPrice;
  String? demands;

  HelpPost(
    String uuid,
    String name,
    String phone,
    String releaseTime, {
    String? expectedAddr,
    int? expectedPrice,
    String? demands,
  })  : expectedAddr = expectedAddr,
        expectedPrice = expectedPrice,
        demands = demands,
        super(uuid, name, phone, releaseTime);

  @override
  Widget buildPanel(BuildContext context,
      {required void Function()? onVerified}) {
    return Padding(
      padding: EdgeInsets.all(8),
      child: Card(
        elevation: 0,
        child: Column(
          children: [
            Padding(
              padding: EdgeInsets.symmetric(vertical: 5),
              child: Text('姓名: $name'),
            ),
            Padding(
              padding: EdgeInsets.symmetric(vertical: 5),
              child: Text('手机: $phone'),
            ),
            Padding(
              padding: EdgeInsets.symmetric(vertical: 5),
              child: Text('期望住址: $expectedAddr'),
            ),
            Padding(
              padding: EdgeInsets.symmetric(vertical: 5),
              child: Text('期望租金: $expectedPrice'),
            ),
            Padding(
              padding: EdgeInsets.symmetric(vertical: 5),
              child: Text('其他需求: $demands'),
            ),
            Padding(
              padding: EdgeInsets.symmetric(vertical: 5),
              child: Text('发布时间: $releaseTime'),
            ),
            Padding(
              padding: EdgeInsets.only(left: 0, top: 0, right: 5, bottom: 5),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: [
                  MaterialButton(
                    onPressed: onVerified,
                    shape: StadiumBorder(),
                    color: Color(0xFFfb966e),
                    elevation: 0,
                    child: Text('审核通过'),
                  )
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class RentPost extends Post {
  String? roomAddr;
  int? roomArea;
  String? roomType;
  String? roomOrientation;
  int? roomFloor;
  String? description;
  int? price;
  String? restriction;
  List<Uint8List>? pictures;

  RentPost(
    String uuid,
    String name,
    String phone,
    String releaseTime, {
    String? roomAddr,
    String? description,
    int? price,
    String? restriction,
    String? roomType,
    int? roomArea,
    int? roomFloor,
    String? roomOrientation,
    List<Uint8List>? pictures,
  })  : pictures = pictures,
        roomAddr = roomAddr,
        description = description,
        price = price,
        restriction = restriction,
        roomType = roomType,
        roomArea = roomArea,
        roomFloor = roomFloor,
        roomOrientation = roomOrientation,
        super(uuid, name, phone, releaseTime);

  @override
  Widget buildPanel(BuildContext context,
      {required void Function()? onVerified}) {
    return Padding(
      padding: EdgeInsets.all(8),
      child: Card(
        elevation: 0,
        child: Column(
          children: [
            Padding(
              padding: EdgeInsets.only(left: 0, top: 0, right: 5, bottom: 5),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.end,
                children: [
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('姓名: $name'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('手机: $phone'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('地址: $roomAddr'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('面积: $roomArea'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('室型: $roomType'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('朝向: $roomOrientation'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('楼层: $roomFloor'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('月租金: $price'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('支付限制: $restriction'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('其他描述: $description'),
                  ),
                  Padding(
                    padding: EdgeInsets.symmetric(vertical: 5),
                    child: Text('发布时间: $releaseTime'),
                  ),
                  MaterialButton(
                    onPressed: onVerified,
                    shape: StadiumBorder(),
                    color: Color(0xFFfb966e),
                    elevation: 0,
                    child: Text('认证通过'),
                  )
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}
